tests/measurements/test_measurements.py

```diff
@@ -430,3 +433,168 @@ def test_obs_diagonalizing_gates(self):
         expected_classes = [qml.PauliZ, qml.S, qml.Hadamard]
         for op, c in zip(res, expected_classes):
             assert isinstance(op, c)
+
+
+@pytest.mark.parametrize("switch_return", [qml.enable_return, qml.disable_return])
+class TestSampleMeasurement:
+    """Tests for the SampleMeasurement class."""
+
+    def test_custom_sample_measurement(self, switch_return):
+        """Test the execution of a custom sampled measurement."""
+        switch_return()
+
+        class MyMeasurement(SampleMeasurement):
+            def process_samples(self, samples, wire_order, shot_range, bin_size):
+                return qml.math.sum(samples[..., self.wires])
+
+        dev = qml.device("default.qubit", wires=2, shots=1000)
+
+        @qml.qnode(dev)
+        def circuit():
+            qml.PauliX(0)
+            return MyMeasurement(wires=[0]), MyMeasurement(wires=[1])
+
+        assert qml.math.allequal(circuit(), [1000, 0])
+
+    def test_sample_measurement_without_shots(self, switch_return):
+        """Test that executing a sampled measurement with ``shots=None`` raises an error."""
+        switch_return()
+
+        class MyMeasurement(SampleMeasurement):
+            def process_samples(self, samples, wire_order, shot_range, bin_size):
+                return qml.math.sum(samples[..., self.wires])
+
+        dev = qml.device("default.qubit", wires=2)
+
+        @qml.qnode(dev)
+        def circuit():
+            qml.PauliX(0)
+            return MyMeasurement(wires=[0]), MyMeasurement(wires=[1])
+
+        with pytest.raises(
+            ValueError, match="Shots must be specified in the device to compute the measurement "
+        ):
+            circuit()
+
+    def test_method_overriden_by_device(self, switch_return):
+        """Test that the device can override a measurement process."""
+        switch_return()
+
+        class MyMeasurement(SampleMeasurement):
+            method_name = "test_method"
+
+            def process_samples(self, samples, wire_order, shot_range, bin_size):
+                return qml.math.sum(samples[..., self.wires])
+
+        dev = qml.device("default.qubit", wires=2, shots=1000)
+
+        @qml.qnode(dev)
+        def circuit():
+            qml.PauliX(0)
+            return MyMeasurement(wires=[0]), MyMeasurement(wires=[1])
+
+        circuit.device.test_method = lambda obs, shot_range=None, bin_size=None: 2
+
+        assert qml.math.allequal(circuit(), [2, 2])
```

Comment:
Ahh I see, so if we initialize a measurement process with `method_name` then if the device implements that method with the same name, then this is called instead of the `process_samples` or `process_state` method we have implemented.

Reply:
Exactly!